### PR TITLE
feat(functions): support custom domains

### PR DESCRIPTION
**Pull request by dackers86**
### Description

Native Sdks have been updated to allow custom domains to be passed when calling functions. This update allows RNF to allow either custom domains or a region to be provided

### Related issues
Fixes https://github.com/invertase/react-native-firebase/issues/4919

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [ ] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [x] Yes
  - [ ] No

BREAKING CHANGE:  SDK min required version bump is required for this feature.

### Test Plan

---

Think `react-native-firebase` is great? Please consider supporting the project with any of the below:

- 👉 Star this repo on GitHub ⭐️
- 👉 Follow [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter

🔥

### CODE REVIEW

**Comment by mikehardy** [comment]
Release notes - should be breaking, and in commit message min version called out:
- requires firebase-ios-sdk 7.1.0 or greater https://firebase.google.com/support/release-notes/ios#cloud-functions_2
- requires firebase-android-sdk 26.2.0 or greater https://firebase.google.com/support/release-notes/android#functions-client_v19-2-0

suggest this as is terse but complete:
> BREAKING(functions): add custom domains requires firebase-ios-sdk 7.1.0+ / firebase-android-sdk 26.2.0+

This is important because there are definitely some issues outstanding where people have only seen success using 26.0.0 or 26.1.0 so to them this is definitely a breaking change at our level as compile will fail on missing symbols if the SDK is not updated even though it is a minor for the native SDKs themselves

I don't see any Java implementation change but on review of the underlying PR https://github.com/firebase/firebase-android-sdk/pull/2100 it appears the API there was fully backwards compatible - verified working there?

**Comment by mikehardy** [comment]
ios E2E CI currently failing on known issue with pending resolution, reference #4951 

```

[!] CocoaPods could not find compatible versions for pod "FirebaseFirestore":
  In Podfile:
    FirebaseFirestore (from `https://github.com/invertase/firestore-ios-sdk-frameworks.git`)

    RNFBFirestore (from `../../packages/firestore`) was resolved to 10.8.1, which depends on
      Firebase/Firestore (~> 7.6.0) was resolved to 7.6.0, which depends on
        FirebaseFirestore (~> 7.6.0)
error Command failed with exit code 1.
```

I'm on that one next / immediately

**Comment by dackers86** [comment]
@mikehardy Yes, test cases added and working for both Android and iOS. 

Android reference shown here https://firebase.google.com/docs/reference/android/com/google/firebase/functions/FirebaseFunctions#public-static-firebasefunctions-getinstance-string-regionorcustomdomain

**Comment by mikehardy** [comment]
@dackers86 as soon as #4953 looks good to you and passes CI I'll merge it, then once you rebase or merge master in the iOS E2E CI here should pass :crossed_fingers:

**Comment by mikehardy** [comment]
Hmm - the PR Title check failure is unfortunate, I thought a 'BREAKING' would work.

The other alternative - which I worked in the past - is a 'feat' PR title / first line and the *commit* message can have a `BREAKING CHANGE:` note at the end, which is more of a squash-merge maintainer note (must do it during the actual merge process)

For extra developer friendliness I usually put BREAKING in top line too like `feat(function): add custom domain BREAKING requires etc etc`.

example https://github.com/invertase/react-native-firebase/commit/70974d41f857a0f7fc09cb5235856d3748b30117 generated changelog https://github.com/invertase/react-native-firebase/blob/master/packages/analytics/CHANGELOG.md#800-2020-11-10

This seems maybe excessively detail oriented but if not done well there will definitely be issues logged later, best to avoid.

File: packages/functions/e2e/functions.e2e.js
Comment (mikehardy) [inline packages/functions/e2e/functions.e2e.js:55]: can you add the test I reference on slack for useEmulator to make sure that emulator overrides custom URL?

File: packages/functions/e2e/functions.e2e.js
Comment (Salakar) [inline packages/functions/e2e/functions.e2e.js:57]: Nit but this isn't a function for region, so should name it different

File: packages/functions/e2e/functions.e2e.js
Comment (Salakar) [inline packages/functions/e2e/functions.e2e.js:63]: Can we use the instance we created above for this, and also the one below?

Comment (dackers86) [reply packages/functions/e2e/functions.e2e.js:63]: Updated 👍

File: packages/functions/e2e/functions.e2e.js
Comment (dackers86) [reply packages/functions/e2e/functions.e2e.js]: Updated 👍

File: packages/functions/e2e/functions.e2e.js
Comment (dackers86) [reply packages/functions/e2e/functions.e2e.js:55]: Added, I get a `PERMISSION DENIED` locally. Running in CI for testing

File: packages/functions/e2e/functions.e2e.js
Comment (dackers86) [reply packages/functions/e2e/functions.e2e.js]: Updated 👍

File: packages/functions/e2e/functions.e2e.js
Comment (mikehardy) [reply packages/functions/e2e/functions.e2e.js:55]: Hmm - I would be concerned if CI reported a different result than local and in fact they both (ios and android) fail with permission denied in CI as well (which is good because consistency is good)

I think this can only be tested via jest anyway, there is no way to unset emulator usage is there? There is not for auth anyway - sending empty or null in is not documented to clear the value and in practice discards I think

Comment (mikehardy) [reply packages/functions/e2e/functions.e2e.js:55]: Really weird from the android log

```

03-01 11:22:22.915  2720 11416 I MdiSyncModule: flag handling is skipping irrelevant intent.
03-01 11:22:22.954  2720  3098 W Conscrypt: Could not set socket write timeout: java.net.SocketException: Socket closed
03-01 11:22:22.954  2720  3098 W Conscrypt: 	at com.google.android.gms.org.conscrypt.Platform.setSocketWriteTimeout(:com.google.android.gms@201817023@20.18.17 (040800-311416286):2)
03-01 11:22:22.954  2720  3098 W Conscrypt: 	at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.setSoWriteTimeout(:com.google.android.gms@201817023@20.18.17 (040800-311416286):0)
03-01 11:22:22.968  2720 11854 W ModuleInitIntentOp: Dropping unexpected action com.google.android.gms.phenotype.COMMITTED
03-01 11:22:22.971  2720  3098 W Conscrypt: Could not set socket write timeout: java.net.SocketException: Socket closed
03-01 11:22:22.971  2720  3098 W Conscrypt: 	at com.google.android.gms.org.conscrypt.Platform.setSocketWriteTimeout(:com.google.android.gms@201817023@20.18.17 (040800-311416286):2)
03-01 11:22:22.971  2720  3098 W Conscrypt: 	at com.google.android.gms.org.conscrypt.ConscryptFileDescriptorSocket.setSoWriteTimeout(:com.google.android.gms@201817023@20.18.17 (040800-311416286):0)
03-01 11:22:23.037  2720 11854 I ChromeSync: [Sync,SyncIntentOperation] Handling the intent: Intent { act=com.google.android.gms.phenotype.COMMITTED flg=0x10 pkg=com.google.android.gms cmp=com.google.android.gms/.chimera.GmsIntentOperationService (has extras) }.
03-01 11:22:23.082  2720 11854 I MdiSyncModule: flag handling is skipping irrelevant intent.
03-01 11:22:23.105  2720 11679 E MS_RegisterService: Exception during register request.
03-01 11:22:23.105  2720 11679 E MS_RegisterService: ciau: PERMISSION_DENIED: The caller does not have permission
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at ciat.c(:com.google.android.gms@201817023@20.18.17 (040800-311416286):3)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at sdn.a(:com.google.android.gms@201817023@20.18.17 (040800-311416286):48)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at afgm.a(:com.google.android.gms@201817023@20.18.17 (040800-311416286):12)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at com.google.android.libraries.matchstick.net.SilentRegisterIntentOperation.a(:com.google.android.gms@201817023@20.18.17 (040800-311416286):179)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at com.google.android.libraries.matchstick.net.SilentRegisterIntentOperation.f(:com.google.android.gms@201817023@20.18.17 (040800-311416286):38)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at com.google.android.libraries.matchstick.net.SilentRegisterIntentOperation.onHandleIntent(:com.google.android.gms@201817023@20.18.17 (040800-311416286):164)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at com.google.android.chimera.IntentOperation.onHandleIntent(:com.google.android.gms@201817023@20.18.17 (040800-311416286):2)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at qgf.onHandleIntent(:com.google.android.gms@201817023@20.18.17 (040800-311416286):4)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at cui.run(:com.google.android.gms@201817023@20.18.17 (040800-311416286):5)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at cuh.run(:com.google.android.gms@201817023@20.18.17 (040800-311416286):9)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
03-01 11:22:23.105  2720 11679 E MS_RegisterService: 	at java.lang.Thread.run(Thread.java:923)
```

That's the stack that lines up with the timestamps and the string PERMISSION_DENIED, but that is not anything like the stack contents I would expect. I'm searching around :thinking:

Comment (mikehardy) [reply packages/functions/e2e/functions.e2e.js:55]: All the online stuff points to google cloud platform enforcing access permissions on some object and kicking out this inscrutable message when it doesn't have it. https://stackoverflow.com/questions/52248879/cloud-composer-permission-denied-the-caller-does-not-have-permission - not sure how to resolve it, but it does appear (online) to be a firestore-related online permission issue

testing emulator-overrides-custom-domain seems best in jest either way - then no worries about this stuff

**Review by mikehardy — APPROVED** [review]
LFantasticTM

File: packages/functions/e2e/functions.e2e.js
Comment (mikehardy) [reply packages/functions/e2e/functions.e2e.js]: verified

File: packages/functions/e2e/functions.e2e.js
Comment (mikehardy) [reply packages/functions/e2e/functions.e2e.js]: verified

**Review by mikehardy — APPROVED** [review]
Just re-scanned with special attention to @Salakar comments to verify they were all fixed. This still LGTM and other comments are verified addressed. I'm running tests locally and checking CI now then I'll release this. Chapeau!

**Comment by mikehardy** [comment]
Yes! The fancy commit message worked: https://github.com/invertase/react-native-firebase/blob/master/CHANGELOG.md#breaking-changes

correctly versioned it as a major, and got the changelog I thought it would.

**Pull request merged**
